File: leetcode/Math/43.MultiplyStrings/43MultiplyStrings.py
'''
max_len = len(num1) + len(num2)
'''
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        if not len(num1) or not len(num2):
            return '0'
        max_len = len(num1) + len(num2)
        res = [0] * max_len
        for i in range(len(num1)-1, -1, -1):
            for j in range(len(num2)-1, -1, -1):
                mul = int(num1[i]) * int(num2[j])
                Pos_low = i+j+1
                Pos_high = i+j
                mul += res[Pos_low]  # 加上之前的计算结果
                res[Pos_low] = mul%10
                res[Pos_high] += mul//10
        res = list(map(str, res))
        str_res = ''.join(res).lstrip('0')
        return str_res if len(str_res) else '0'

File: leetcode/Math/43.MultiplyStrings/43MultiplyStrings.cpp
//高精度乘法题
class Solution {
public:
    string multiply(string num1, string num2) {
        vector<int> product(num1.size() + num2.size(), 0);
        for (int i=0; i<num1.size(); i++)
            for (int j=0; j<num2.size(); j++)
                product[num1.size()-i+num2.size()-j-2] += (num1[i]-'0')*(num2[j]-'0');
        int t = 0;
        for (int i=0; i<product.size(); i++)
        {
            int &x = product[i];
            t += x;
            x = t % 10;
            t /= 10;
        }
        string res;
        int k = product.size()-1;
        while (!product[k] && k>0) k--;
        for (int i=k; i>=0; i--) res += to_string(product[i]);
        return res;
    }
};